### ISSUE TITLE
Question on the gradient

### ISSUE THREAD

**Issue by kishorkl** [GRAAL-Research/domain_adversarial_neural_network#8]
I recently was trying to simulate the results from the paper, for the toy example with the two moons,  and stumbled up on this code. Thanks for the code, but I feel that the gradient being used here has a term missing.

Using the same labels as per the code, following are the networks in the code:

hidden_layer = sigmoid(W.X + b)
output = softmax(V.hidden_layer + c)
domain = sigmoid(U.hidden_layer + d)

The cost function(J) for the domain is D - D-hat (assuming D is the true domain and D-hat is the predicted). Hence the gradient of the domain-loss with respect to the bias (d) of the domain-network should be: (read delta-x/delta-y as dx/dy)

delta-J/delta-d = delta-J/delta-sigmoid * delta-sigmoid / delta-(U.hidden_layer + d) * delta-(U.hidden_layer + d) / delta-d
                         = -1 * sigmoid . (1 - sigmoid) * 1

And hence the update of the 'd' bias should be: (I used the derivative of sigma function from (https://towardsdatascience.com/derivative-of-the-sigmoid-function-536880cf918e ) 

d - lr * delta_d = d - lr * (-1 * sigmoid . (1 - sigmoid))
But as per the code, the update term is 

d - lr * delta_d = d - lr * (domain - est-domain) 

this is extended for both source domain and target domain.

i do not see the derivative of sigmoid being used here. Is there any explanation for this?

Thanks a lot in advance and merry Christmas,
Kishor Kayyar

**Comment by kishorkl** [comment]
After some more analysis and re-reading of the paper I realize that the DA training works if cross entropy loss (instead of L1 loss as assumed in the earlier comment) is used for domain classifier. The implementation is also correct in this case..
